src/main/scala/com/microsoft/hyperspace/index/execution/BucketUnionExec.scala

```diff
@@ -0,0 +1,114 @@
+ * retain outputPartitioning of result. So we define a new Union operation complying
+ * with the following conditions:
+ *   - input RDDs must have the same number of partitions.
+ *   - input RDDs must have the same partitioning keys.
+ *   - input RDDs must have the same column schema.
+ *
+ * Unfortunately, there is no explicit API to check Partitioning keys in RDD, we have to
+ * assure that on the caller side. Therefore, [[BucketUnionRDD]] is Hyperspace
+ * internal use only.
+ *
+ * You can find more detailed information about Bucketing optimization in
+ * [[https://youtu.be/7cvaH33S7uc Bucketing 2.0: Improve Spark SQL Performance by Removing Shuffle]]
+ */
+private[hyperspace] class BucketUnionRDD[T: ClassTag](
+    sc: SparkContext,
+    var rdds: Seq[RDD[T]],
+    bucketSpec: BucketSpec)
+    extends RDD[T](sc, rdds.map(x => new OneToOneDependency(x))) {
+  require(rdds.nonEmpty)
+  require(rdds.forall(_.getNumPartitions == bucketSpec.numBuckets))
+
+  // copy from org.apache.spark.rdd.PartitionerAwareUnionRDD
+  override def getPartitions: Array[Partition] = {
+    val numBuckets = bucketSpec.numBuckets
+    (0 until numBuckets).map { index =>
+      new BucketUnionRDDPartition(rdds, index)
+    }.toArray
+  }
+
+  // copy from org.apache.spark.rdd.PartitionerAwareUnionRDD
+  override def compute(s: Partition, context: TaskContext): Iterator[T] = {
+    val parentPartitions = s.asInstanceOf[BucketUnionRDDPartition].parents
+    rdds.zip(parentPartitions).iterator.flatMap {
+      case (rdd, p) => rdd.iterator(p, context)
+    }
+  }
+
+  // copy from org.apache.spark.rdd.PartitionerAwareUnionRDD
+  override def clearDependencies() {
+    super.clearDependencies()
+    rdds = null
+  }
+}
+
+/**
+ * [[BucketUnionRDDPartition]] keeps partitions for each partition index.
+ * @param rdds  Input RDDs.
+ * @param index Partition index.
+ */
+private[hyperspace] class BucketUnionRDDPartition(
+    @transient val rdds: Seq[RDD[_]],
+    override val index: Int)
+    extends Partition {
+  val parents: Array[Partition] = rdds.map(_.partitions(index)).toArray
+
+  override def hashCode(): Int = index
+  override def equals(other: Any): Boolean = super.equals(other)
+}
+
+/**
+ * [[BucketUnionExec]] is Spark Plan for [[BucketUnion]].
+ *
+ * @param children Child plans.
+ * @param bucketSpec Bucket specification.
+ */
+private[hyperspace] case class BucketUnionExec(children: Seq[SparkPlan], bucketSpec: BucketSpec)
+    extends SparkPlan {
+  override protected def doExecute(): RDD[InternalRow] = {
+    new BucketUnionRDD[InternalRow](sparkContext, children.map(_.execute()), bucketSpec)
+  }
+
+  override def output: Seq[Attribute] = children.head.output
+
+  // Here we could not add some check for outputPartitioning of children because the
+  // child plans do not have the correct outputPartitioning until the generation of a
+  // complete physical plan. Therefore, let [[BucketUnionRDD]] check the number of
+  // partitions after this.
```

Comment:
> It seems this SparkPlan used for some intermediate class(?) in catalyst optimizer.
>
> I printed child.outputPartitioning.numPartitions in require function while an execution of hs.explain(query),
>
> in require()
> 0 0 // this is at the first creation of BucketUnionExec in BucketUnionStrategy
> in require()
> 200 0 1
> in require()
> 200 200
> in require()
> 200 200
> So.. I think the requires in BucketAwareUnionRDD are enough.

I couldn't comment on https://github.com/microsoft/hyperspace/pull/123#discussion_r475366280, but can you try something like the following:
```scala
  override def outputPartitioning: Partitioning = {
    children.map(_.outputPartitioning).foreach {
      case h: HashPartitioning => println(h.numPartitions)
      case other => throw new HyperspaceException("Only HashPartitioning is supported.")
    }
    children.head.outputPartitioning
  }
```
if you see different numbers?

Reply:
It works! I'll add below code 👍 
```
override def outputPartitioning: Partitioning = {
    children.map(_.outputPartitioning).foreach {
      case h: HashPartitioning => println(h.numPartitions)
      case other => throw new HyperspaceException("Only HashPartitioning is supported.")
    }
    children.head.outputPartitioning
  }
```

Reply:
Oh, this was just a debug code. I would check if all the children have the `HashPartitioning` with the same `numPartitions`, and `expressions` match up.

Reply:
Or maybe the following?
```scala
    assert(children.map(_.outputPartitioning).toSet.size == 1)
    assert(children.head.outputPartitioning.isInstanceOf[HashPartitioning])
```

Reply:
I previously checked the numbers in `require`, but it seems `def outputPartitioning` called only with some proper children node. Anyway I added both assert inside `outputPartitioning` here: 
https://github.com/microsoft/hyperspace/pull/151/files#diff-1c44518b3fe9eb338fc8250ec5794b62R110